app/src/Application.php

```diff
@@ -115,6 +124,37 @@ public function services(ContainerInterface $container): void
     {
     }
 
+    /**
+     * Build and return authentication service.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request the auth request
+     * @return \Authentication\AuthenticationServiceInterface
+     */
+    public function getAuthenticationService(ServerRequestInterface $request): AuthenticationServiceInterface
+    {
+        $authenticationService = new AuthenticationService([
+            'unauthenticatedRedirect' => Router::url('/users/login'),
```

Comment:
```suggestion
            'unauthenticatedRedirect' => Router::url(['controller' => 'Users', 'action' => 'login']),
```


Instead of using string URL, we normally used the array format routing. It's more verbose and easy to maintain later (The URL string might be changed but the controller and action rarely change).

Please refer https://book.cakephp.org/4/en/development/routing.html#routing-arrays-vs-request-parameters